### PR TITLE
Add checks to turn off material dependency and derivative material coupling

### PR DESCRIPTION
**Pull request by tophmatthews**
Ref #14914

### CODE REVIEW

**Comment by tophmatthews** [comment]
ping @dschwen, this is based on profiling you did while you were at LANL some time back. This cuts it down from 29 to 13 seconds. The big problem maker in `DerivateMaterialInterface` is `haveMaterialProperty`. Not validating coupling, as done here cuts down how many times `haveMaterialProperty` is called, but it can be cut down by about a half-ish if `haveMaterialProperty` is always set to true, which is a bit hacky. Do you see another way?

**Comment by tophmatthews** [comment]
profile:
dep_val

File: framework/src/materials/DerivativeSumMaterial.C [hunk 1]
```diff
@@ -38,6 +38,8 @@ DerivativeSumMaterial::validParams()
   // Advanced arguments to construct a sum of the form \f$ c+\gamma\sum_iF_i \f$
   params.addParam<std::vector<Real>>("prefactor", "Prefactor to multiply the sum term with.");
   params.addParam<Real>("constant", 0.0, "Constant to be added to the prefactor multiplied sum.");
+
+  params.addParam<bool>("validate_coupling", true, "Flag to validate coupling.");
```
Comment (dschwen) [inline framework/src/materials/DerivativeSumMaterial.C]:
```suggestion
  params.addParam<bool>("validate_coupling", true, "Check if all variables the specified materials depend on are listed in the `args` parameter.");
```

That would make it a bit more descriptive

File: framework/src/materials/DerivativeSumMaterial.C [hunk 2]
```diff
@@ -105,8 +108,9 @@ DerivativeSumMaterial::DerivativeSumMaterial(const InputParameters & parameters)
 void
 DerivativeSumMaterial::initialSetup()
 {
-  for (unsigned int n = 0; n < _num_materials; ++n)
-    validateCoupling<Real>(_sum_materials[n]);
+  if (_validate_coupling)
+    for (unsigned int n = 0; n < _num_materials; ++n)
+      validateCoupling<Real>(_sum_materials[n]);
```
Comment (dschwen) [inline framework/src/materials/DerivativeSumMaterial.C]: :+1: I think this is the best way of doing it. It allows users to set a GlobalParam if they want to skip this check.

**Review by dschwen — DISMISSED** [review]

**Comment by tophmatthews** [comment]
The main problem is `haveMaterialProperty` is used to set a default (rather a zero) material property if it doesn't exist. The only way I can think to speed `haveMaterialProperty` (and really it's `hasBlockMaterialPropertyHelper` is to precompute a table...or something...I dunno.

**Comment by dschwen** [comment]
> `haveMaterialProperty` is used to set a default (rather a zero) material property if it doesn't exist

Huh? That should not happen. Only if you *get* a default material property the zero property should be constructed.

**Comment by tophmatthews** [comment]
Whoops, yeah, it's this line:
```
DerivativeMaterialInterface<T>::getDefaultMaterialPropertyByName(const std::string & prop_name)
{
  // if found return the requested property
  if (haveMaterialProperty<U>(prop_name))
    return getMaterialPropertyByName<U>(prop_name);

  return this->template getZeroMaterialProperty<U>(prop_name);
}
```

**Comment by dschwen** [comment]
Ah, I see what you're saying. That's one case of `haveMaterialProperty` you cannot easily get rid of. :-/ 
What faction of those calls remains?

**Comment by tophmatthews** [comment]
It's about a third of the runtime
image

**Comment by dschwen** [comment]
I meant with your intervention, is it still 1/3?

**Comment by tophmatthews** [comment]
> I meant with your intervention, is it still 1/3?

Yes

**Comment by tophmatthews** [comment]
The performance map above is with the two interventions included here.

File: test/tests/materials/derivative_sum_material/random_ic.i [hunk 3]
```diff
@@ -3,28 +3,15 @@
...
-    family = LAGRANGE
     [./InitialCondition]
-      type = SmoothCircleIC
-      x1 = 125.0
-      y1 = 125.0
-      radius = 80.0
-      invalue = 1.0
-      outvalue = 0.1
-      int_width = 80.0
+      type = RandomIC
```
Comment (tophmatthews) [inline test/tests/materials/derivative_sum_material/random_ic.i]: FYI @dschwen  I pulled this test out of phase_field by changing this IC. I think this still tests the DerivativeSumMaterial in the same way, and it keeps the test with the src.

File: framework/include/problems/FEProblemBase.h [hunk 4]
```diff
@@ -1978,6 +1978,9 @@ class FEProblemBase : public SubProblem, public Restartable
   /// Determines whether a check to verify an active material on every subdomain
   bool _material_coverage_check;
 
+  /// Determines whether a check to verify material dependencies on every subdomain
+  bool _material_dependency_check;
```
Comment (dschwen) [inline framework/include/problems/FEProblemBase.h]: you could make this `const` as well

**Review by dschwen — DISMISSED** [review]

File: framework/include/problems/FEProblemBase.h (hunk 4 above)
Comment (tophmatthews) [reply framework/include/problems/FEProblemBase.h]: AH! yes of course.

**Comment by tophmatthews** [comment]
Hopefully the last change @dschwen, added const

**Review by dschwen — DISMISSED** [review]

**Comment by tophmatthews** [comment]
this should be good to go @permcody

**Comment by permcody** [comment]
I'm assigning to @dschwen for final approval. It looks like he's got it under control.

**Review by dschwen — APPROVED** [review]

**Pull request merged**